workloads/librispeech/librispeech_pytorch/workload.py

```diff
@@ -0,0 +1,219 @@
+"""LibriSpeech workload implemented in Pytorch."""
+
+import itertools
+import os
+from typing import Tuple
+
+from . import input_pipeline
+from . import models
+import ctcdecode
+import Levenshtein
+import spec
+import torch
+
+
+class LibriSpeechWorkload(spec.Workload):
+  """A LibriSpeech workload."""
+
+  def __init__(self):
+    self._train_loader = None
+    self._valid_loader = None
+    self._device = torch.device(
+        "cuda:0" if torch.cuda.is_available() else "cpu")
+    self._loss = torch.nn.CTCLoss(blank=0, reduction="none")
+    self._label_dict = {
+        "_": 0,
+        " ": 1,
+        "'": 2,
+        "A": 3,
+        "B": 4,
+        "C": 5,
+        "D": 6,
+        "E": 7,
+        "F": 8,
+        "G": 9,
+        "H": 10,
+        "I": 11,
+        "J": 12,
+        "K": 13,
+        "L": 14,
+        "M": 15,
+        "N": 16,
+        "O": 17,
+        "P": 18,
+        "Q": 19,
+        "R": 20,
+        "S": 21,
+        "T": 22,
+        "U": 23,
+        "V": 24,
+        "W": 25,
+        "X": 26,
+        "Y": 27,
+        "Z": 28,
+    }
+    self._rev_label_dict = {v: k for k, v in self._label_dict.items()}
+    self._decoder = ctcdecode.CTCBeamDecoder(
+        labels=[str(c) for c in self._rev_label_dict], beam_width=1)
+
+  def has_reached_goal(self, eval_result: float) -> bool:
+    return eval_result < self.target_value
+
+  def build_input_queue(self, data_rng, split: str, data_dir: str,
+                        batch_size: int):
+    torch.manual_seed(data_rng[0])
+    train_set = input_pipeline.LibriSpeechDataset(
+        os.path.join(data_dir, "features_train-clean-100.csv"))
+    valid_set = input_pipeline.LibriSpeechDataset(
+        os.path.join(data_dir, "features_test-clean.csv"))
+
+    train_collate_fn = train_set.pad_collate
+
+    self._train_loader = torch.utils.data.DataLoader(
+        train_set,
+        batch_size=batch_size,
+        shuffle=True,
+        pin_memory=True,
+        collate_fn=train_collate_fn)
+
+    self._valid_loader = torch.utils.data.DataLoader(
+        valid_set,
+        batch_size=batch_size,
+        pin_memory=True,
+        collate_fn=train_collate_fn)
```

Comment:
Potentially it could be beneficial to set the argument `num_workers` of the `DataLoader` classes to a positive integer (the default is `0`) -- see [here](https://pytorch.org/docs/stable/data.html#single-and-multi-process-data-loading) for the docs. It's not clear how to choose the optimal value, it might best to increase it by `+1` until the training process is not getting faster anymore. It might also not help at all, but it could be worth trying if it hasn't been considered yet.

Reply:
We will use single host rather than multi host training, so we don't have to set num_workers?

Reply:
Sorry, I don't understand what you mean with "single host training". Do you mean that the entire dataset can be loaded in memory (because it's small enough), or that only a single CPU core is used (because one [suggestion](https://pytorch-lightning.readthedocs.io/en/latest/guides/speed.html#num-workers) is to use `num_workers=os.cpu_count()`)? In case you mean that we are only using data parallelism on a single node with multiple GPUs, setting `num_workers > 0` might still help if I'm not mistaken (assuming that the whole dataset cannot be loaded in memory and that there is more than one CPU core).

Reply:
I set num_workers=2 after some trials.